interfaces/hotplug/proposed_slot.go

```diff
@@ -47,43 +44,25 @@ type RequestedSlotSpec struct {
 	Attrs map[string]interface{} `json:"attrs,omitempty"`
 }
 
-// Specification contains a slot definition to create in response to a hotplug event.
-type Specification struct {
-	slot *RequestedSlotSpec
-}
-
-// NewSpecification creates an empty hotplug Specification.
-func NewSpecification() *Specification {
-	return &Specification{}
-}
-
-// SetSlot sets the slot specification.
-func (h *Specification) SetSlot(slotSpec *RequestedSlotSpec) error {
-	if h.slot != nil {
-		return fmt.Errorf("slot specification already created")
-	}
+// Clean returns a copy of the input slot with normalized attributes and validated slot name (unless its empty).
+func (slot *ProposedSlot) Clean() (*ProposedSlot, error) {
 	// only validate name if not empty, otherwise name is created by hotplug
-	// subsystem later on when the spec is processed.
-	if slotSpec.Name != "" {
-		if err := snap.ValidateSlotName(slotSpec.Name); err != nil {
-			return err
+	// subsystem later on when the proposed slot is processed.
+	if slot.Name != "" {
+		if err := snap.ValidateSlotName(slot.Name); err != nil {
+			return nil, err
 		}
 	}
-	attrs := slotSpec.Attrs
+	attrs := slot.Attrs
 	if attrs == nil {
 		attrs = make(map[string]interface{})
 	} else {
-		attrs = utils.CopyAttributes(slotSpec.Attrs)
-	}
-	h.slot = &RequestedSlotSpec{
-		Name:  slotSpec.Name,
-		Label: slotSpec.Label,
-		Attrs: utils.NormalizeInterfaceAttributes(attrs).(map[string]interface{}),
+		attrs = utils.CopyAttributes(slot.Attrs)
```

Comment:
doesn't NormalizeInterfaceAttributes essentially do a copy too? do we need this copy?

Reply:
Ah, of course, you're right, fixed. Thanks!